### PR TITLE
Challenge submission

### PR DESCRIPTION
**Pull request by kathicks**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,10 +1,22 @@
...
+gem 'rspec'                       # unit testing library
+gem 'rspec-sinatra'               # addon unit testing library for sinatraß
+gem 'capybara'                    # feature testing for web applications
+gem 'cucumber'
+gem 'rake'
+gem 'coveralls', require: false   # analysing test coverage
+gem 'rubocop-rspec'               # analysing test cleanliness/style
+gem 'rubocop'                     # analysing code cleanliness/style
+gem 'database_cleaner'
+gem 'timecop'
```
Comment (KatHicks) [inline Gemfile]: Used the timecop gem for stubbing time in Capybara. Do you know if this is the best way to go about this?

File: Rakefile [hunk 2]
```diff
@@ -1,7 +1,25 @@
+require 'data_mapper'
```
Comment (KatHicks) [inline Rakefile]: Still don't really understand the need for rake. Are there any good resources online that would help me better understand it?

File: app/app.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
+ENV["RACK_ENV"] ||= "development"
```
Comment (KatHicks) [inline app/app.rb]: Likewise to rake, are there any good resources for understanding rack? I've got the basic idea, but still don't fully understand how it is working.

File: app/app.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+require 'sinatra/base'
+require 'sinatra/flash'
+
+require_relative 'datamapper_setup'
+
+class Chitter < Sinatra::Base
```
Comment (KatHicks) [inline app/app.rb]: Now that we've moved onto more complex database apps, it feels like the cleanliness of my code and separation of concerns has got a bit lost. Is this controller okay in terms of design and slimness? What would be the best and highest priority ways of refactoring my controller here?

File: app/public/bootstrap-3.3.7/js/npm.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// This file is autogenerated via the `commonjs` Grunt task. You can require() this file in a CommonJS environment.
```
Comment (KatHicks) [inline app/public/bootstrap-3.3.7/js/npm.js]: It's interesting that bootstrap doesn't pass hound. Does this matter?

File: app/public/stylesheet.css [hunk 6]
```diff
@@ -0,0 +1,109 @@
+.banner {
```
Comment (KatHicks) [inline app/public/stylesheet.css]: I had a lot of fun with the styling but still not quite finished - and annoys me that a few elements are misaligned or unstyled in the app. Also think that this file is pretty messing - want to tidy it up a bit.

File: app/views/index.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+<div class='container-fluid highlights'>
+  <h3 class='subtitle'> Moments | <a href='/login'>View all</a></h3>
+  <div class='row'>
+    <% @peeps.each do |peep| %>
+        <div class="col-md-3 col-xs-6">
+          <div class='sample'>
+            <p class='content'><%= peep.content %></p>
+            <p class='author'><b><%= peep.user.name %></b> <em>@<%= peep.user.username %></em></p>
+            <p class='footnote'>Posted on <%= peep.created_at.strftime('%d %B %Y at %H:%M') %></p>
```
Comment (KatHicks) [inline app/views/index.erb]: Should this be extracted to the Peep model?

Comment (KatHicks) [inline app/views/index.erb]: How do you refer to properties in methods within DataMapper classes? Is it the same as referring to instance variables?

File: app/views/layout.erb [hunk 8]
```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE html>
+<html>
+
+  <head>
```
Comment (KatHicks) [inline app/views/layout.erb]: Layouts are really helpful! Big improvement on my challenge from last week.

File: app/views/peeps.erb [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<div class='posting'>
+  <h2>What's happening?</h2>
+  <form id='peeping' action='/peeps/new' method='post'>
+    <textarea class="message form-control" form='peeping' rows="2" maxlength="140" name='peep' required></textarea>
+    <br>
+    <input class="btn btn-default navbar-btn" type='submit' value='Peep'>
+    <span class="countdown"></span>
```
Comment (KatHicks) [inline app/views/peeps.erb]: I'd been hoping to add a character countdown to the textarea box (as is the case in the real Twitter) - but did a bit of research and realised that it needed JQuery and never got round to it. If time had not been a limiting factor, I'd have liked to figure this out.

File: app/views/users/new.erb [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  <label for='password' class='sr-only'></label>
+  <input placeholder="Password" type='text' name='password' id='password' class="form-control">
+
+  <label for='confirm_password' class='sr-only'></label>
+  <input placeholder="Confirm password" type='text' name='confirm_password' id='confirm_password' class="form-control" >
+
+  <input class="btn btn-lg btn-primary btn-block" type='submit' value='Submit'>
+</form>
+
+<% if flash[:errors] %>
```
Comment (KatHicks) [inline app/views/users/new.erb]: Is there a way of styling flash errors? Had a go in the core stylesheet file - but it didn't work. Seems that maybe you need to make a separate stylesheet for flash errors but again, didn't have time to properly look into it.

File: spec/features/1_user_signup_spec.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+feature '1: user sign up' do
```
Comment (KatHicks) [inline spec/features/1_user_signup_spec.rb]: I've not included any unit tests in my application. We weren't shown how to test DataMapper models (as opposed to pure Ruby classes) in the week's challenge - and wasn't sure how to approach it. Do you have any hints or tips?

File: app/views/users/new.erb [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  <div id="errors">
+    <% flash[:errors].each do |error| %>
+      <div class='error'><p><%= error %></p></div>
+    <% end %>
+  </div>
```
Comment (neoeno) [inline app/views/users/new.erb]: I'm being pernickety here but it's usually a good idea to have errors appear above the form. Apart from it being more noticeable to most people, it also helps visually impaired people. If you use a screen reader you experience the page linearly, so until you get all the way beneath the form it looks like exactly the same page — harder to notice the errors!

**Pull request closed without merging**